### ISSUE TITLE
the pointer triggered by WidgetTester.longPress did not last as long as expected

### ISSUE THREAD

**Issue by xhzq233** [flutter/flutter#129984]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

use `flutter test` to run code example

### Expected results

`onPointerUpTime - onPointerDownTime` is equal to `(kLongPressTimeout + kPressTimeout).inMilliseconds`

### Actual results

`onPointerUpTime - onPointerDownTime` is around 45. However, expect duration is 600.

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

int get now => DateTime.now().millisecondsSinceEpoch;

int? onPointerDownTime;
int? onPointerUpTime;

void main() {
  testWidgets('Tap test', (WidgetTester tester) async {
    await tester.pumpWidget(Listener(
      key: const Key('listener'),
      onPointerDown: (event) {
        print('onPointerDown time: $now');
        onPointerDownTime = now;
      },
      onPointerUp: (_) {
        print('onPointerUp time: $now');
        onPointerUpTime = now;
      },
      child: const SizedBox(width: 100, height: 100, child: const ColoredBox(color: Colors.blue)),
    ));

    final listener = find.byKey(const Key('listener'));
    expect(listener, findsOneWidget);
    // Dispatch a pointer down / pointer up sequence (with a delay of
    // [kLongPressTimeout] + [kPressTimeout] between the two events)
    await tester.longPress(listener);
    await tester.pump();

    print('onPointerUpTime - onPointerDownTime: ${onPointerUpTime! - onPointerDownTime!}');
    final expectDuration = (kLongPressTimeout + kPressTimeout).inMilliseconds;
    print('expectDuration: $expectDuration');
    assert(onPointerUpTime! - onPointerDownTime! == expectDuration);
  });
}

```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.4, on macOS 13.4.1 22F82 darwin-arm64, locale en-JP)
    • Flutter version 3.10.4 on channel stable at /Users/xhz/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 682aa387cf (4 weeks ago), 2023-06-05 18:04:56 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1
    • Pub download mirror https://mirrors.tuna.tsinghua.edu.cn/dart-pub

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/xhz/Library/Android/sdk
    • Platform android-33, build-tools 33.0.2
    • ANDROID_HOME = /Users/xhz/Library/Android/sdk
    • Java binary at: /Users/xhz/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/xhz/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.3)
    • IntelliJ at /Users/xhz/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 74.0.4
    • Dart plugin version 231.9161.14

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.3)
    • IntelliJ at /Users/xhz/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9161.38/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.79.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop)           • macos                     • darwin-arm64   • macOS 13.4.1 22F82 darwin-arm64
    • Chrome (web)              • chrome                    • web-javascript • Google Chrome 114.0.5735.198
```

</details>

**Comment by huycozy** [comment]
Hi @xhzq233 
I'm quite confused about the Expected result. The delta time value between pointer down and up differs between test time, so it can't be equal to 600. I wonder if this value should be a constant or if it has already been defined somewhere that I don't know. Can you please be more specific?

Also, I found there is an issue that could be similar to this (https://github.com/flutter/flutter/issues/100592), please take a look at it and confirm if it's the case you are also having now.

**Comment by xhzq233** [comment]
Even if this value should not be equal to 600, it should be around 600, not around 45. The 600 is the result calculated by kLongPressTimeout + kPressTimeout.
see https://github.com/flutter/flutter/blob/5b12b7467fcbbdc7351d76690ce8a8693e804179/packages/flutter_test/lib/src/controller.dart#L602C3-L608C4
Is the `pump` function not running as expected?

**Comment by xhzq233** [comment]
Sorry, I closed it by mistake.

**Comment by huycozy** [comment]
Thanks for your response. It could be an issue of `pump`. Also, there is a similar open issue (https://github.com/flutter/flutter/issues/95203) but it's caused by `runAsync`. It is possible that this issue is the cause or closely related to that issue.

The output log corresponds to the given sample code:
```console
00:25 +0: Tap test                                                                                                         
onPointerDown time: 1688616435296
onPointerUp time: 1688616435407
onPointerUpTime - onPointerDownTime: 109
expectDuration: 600
```

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.10.5, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.10.5 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (3 weeks ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.79.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.66.0

[✓] Connected device (4 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D                         • android-arm64  • Android 11 (API 30)
    • iPhone (mobile)  • d9a94afe2b649fef56ba0bfeb052f0f2a7dae95e • ios            • iOS 15.7.2 19H218
    • macOS (desktop)  • macos                                    • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)     • chrome                                   • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.12.0-14.0.pre.48, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.12.0-14.0.pre.48 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision bc49cd1bca (3 hours ago), 2023-07-05 17:42:57 -0700
    • Engine revision bd2e42b203
    • Dart version 3.1.0 (build 3.1.0-262.0.dev)
    • DevTools version 2.25.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.79.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.68.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by goderbauer** [comment]
cc @pdblasi-google

**Comment by pdblasi-google** [comment]
So, this is a side effect of how `flutter_test` is designed.

**TL;DR** This is expected, due to [dart-lang #28985](https://github.com/dart-lang/sdk/issues/28985).

When tests are run using `flutter_test`, they are actually run under a `FakeAsync` environment, which allows things like waits, pumps, and time-based interactions (like long-presses) to happen without requiring the full time to elapse in the test run. Because of the issue mentioned above, this comes with the side effect of the global clock for Flutter being out of sync with `DateTime.now()`. There are a couple ways you can work around this.

First, you can use `clock.now()` instead of `DateTime.now()`. If you update your test to include `package:clock/clock.dart`, and change your `now` getter to `int get now => clock.now().millisecondsSinceEpoch`, you'll get the results you're expecting. This has the benefit of working for both `flutter_test` and `integration_test`.

Second, `DateTime.now()` should work as you expect here when you run your test using `integration_test` instead of `flutter_test` since that doesn't run in a `FakeAsync` environment.
